common/scala/src/main/scala/org/apache/openwhisk/core/containerpool/logging/DockerToActivationLogStore.scala

```diff
@@ -87,16 +97,18 @@ class DockerToActivationLogStore(system: ActorSystem) extends LogStore {
    */
   protected def logStream(transid: TransactionId,
                           container: Container,
-                          action: ExecutableWhiskAction,
-                          isTimedoutActivation: Boolean): Source[ByteString, Any] = {
-
-    // wait for a sentinel only if no container (developer) error occurred to avoid
-    // that log collection continues if the action code still logs after timeout
-    val sentinel = action.exec.sentinelledLogs && !isTimedoutActivation
-    val logs = container.logs(action.limits.logs.asMegaBytes, sentinel)(transid)
-    val logsWithPossibleError = if (isTimedoutActivation) {
+                          logLimit: LogLimit,
+                          sentinelledLogs: Boolean,
+                          isDeveloperError: Boolean): Source[ByteString, Any] = {
+
+    // Wait for a sentinel only if no container (developer) error occurred to avoid
+    // that log collection continues if the action code still logs after developer error.
+    val waitForSentinel = sentinelledLogs && !isDeveloperError
+    val logs = container.logs(logLimit.asMegaBytes, waitForSentinel)(transid)
+    val logsWithPossibleError = if (isDeveloperError) {
       logs.concat(
-        Source.single(ByteString(LogLine(Instant.now.toString, "stderr", Messages.logFailure).toJson.compactPrint)))
+        Source.single(
+          ByteString(LogLine(Instant.now.toString, "stderr", Messages.logWarningDeveloperError).toJson.compactPrint)))
```

Comment:
I'm assuming the log will end with both a logging error and this error, is that correct? If so: maybe we should drop the TimeoutException from the recover below and add that to the the stream here. We can then decide if we want to append the general log failure or this new one.

Reply:
> I'm assuming the log will end with both a logging error and this error, is that correct?

I think no, because:

* If we have a developer error, we won't perform tailing log file reading but only read to the file's current end. In addition, we do not require to find 2 sentinel markers.
* With these conditions, the likelihood to run into idle / completion timeouts should be low.
* All other errors (framing, log limit exceeded) could possibly occur - but there is no correlation to the developer error.

My understanding is that you are thinking about breaking ties when we have two different error conditions like developer error AND timeout / framing error / log limit exceeded?

From my point of view, there are three perspectives on that:

1. As action developer, I use the action logs to understand and debug problems with init'ing and running actions. For the action developer, it probably doesn't matter which error appears in the last log line or whether the two last log lines mention more specific error conditions (`There was an issue ...` and `The action did not initialize or run as expected...`) - as long as there is a hint what happened and that the log is probably truncated.

2. From a container proxy perspective, I want to remove (and not re-use) an action container that had a developer or log collecting and processing error. From that perspective, it doesn't matter which of these two error conditions arrives in the container proxy.

3. As operator, I do not want see indications (system log messages, metrics) for failed log collections if there is a developer error - but log collecting and processing works properly. Only if a timeout / framing error / log limit exceeded occurs, I want a log collecting failure. That's why I excluded `logWarningDeveloperError` as possible error message.

The current variant would not "see" any of the current error messages `There was an issue while collecting your logs. Data might be missing.` and `Logs were truncated because the total bytes size exceeds the limit of ${limit.toBytes} bytes.` in case of a developer error because the last message would always be `The action did not initialize or run as expected. Log data might be missing.`. From that perspective, it would be good append the developer error message after checking the last log line for log collecting / processing errors.

Reply:
Thanks for the exhaustive writeup. My only concern was having multiple error messages in the log file. The missing piece in my head was that we don't tail in the non-sentinel case (I'm hitting my past self for this, I should have introduced a separate path for that which doesn't have the idleTimeout for instance 😂 )

Reply:
@markusthoemmes I pushed a new commit that should properly deal with log collecting errors when there is a developer error. Please have a look. It's no "nice" code though...